src/s3/worker.c

```diff
@@ -89,6 +133,148 @@ register_s3worker(int num)
+ */
+static void
+s3_workers_wait_for_flush(void)
+{
+	Assert(workers_ctl != NULL);
+
+	for (;;)
+	{
+		int			all_flushed = pg_atomic_read_u32(&workers_ctl->pgFilesCnt) == 0;
+
+		for (int i = 0; (i < s3_num_workers) && all_flushed; i++)
+		{
+			if (!pg_atomic_unlocked_test_flag(&workers_ctl->workersInProgress[i]))
+				all_flushed = false;
+		}
+		if (all_flushed)
+			break;
+
+		ConditionVariableTimedSleep(&workers_ctl->pgFilesFlushedCV, BgWriterDelay,
+									WAIT_EVENT_BGWRITER_MAIN);
+	}
+
+	ConditionVariableCancelSleep();
+}
+
+/*
+ * Prepare S3 workers to checkpoint PGDATA files.
+ */
+void
+s3_workers_checkpoint_init(void)
+{
+	/* Just in case delete any leftover files */
+	for (int i = 0; i < s3_num_workers; i++)
+	{
+		char		worker_filename[MAXPGPATH];
+
+		pg_atomic_clear_flag(&workers_ctl->workersInProgress[i]);
+
+		snprintf(worker_filename, sizeof(worker_filename), "%s.%d",
+				 PGFILES_CRC_FILENAME, i);
+
+		unlink(worker_filename);
+	}
+}
+
+/*
+ * Compact all S3 workers checksum files into one file.
+ */
+void
+s3_workers_checkpoint_finish(void)
+{
+	int			file;
+
+	s3_workers_wait_for_flush();
+
+	file = BasicOpenFile(PGFILES_CRC_FILENAME, O_CREAT | O_WRONLY | O_TRUNC | PG_BINARY);
+	if (file < 0)
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not create file \"%s\": %m", PGFILES_CRC_FILENAME)));
+
+	for (int i = 0; i < s3_num_workers; i++)
+	{
+		int			worker_file;
+		char		worker_filename[MAXPGPATH];
+		Size		readBytes;
+		char		buffer[8192];
+
+		snprintf(worker_filename, sizeof(worker_filename), "%s.%d",
+				 PGFILES_CRC_FILENAME, i);
+
+		worker_file = BasicOpenFile(worker_filename, O_RDONLY | PG_BINARY);
+		if (worker_file < 0)
+		{
+			/*
+			 * In case if this worker didn't manage to process any
+			 * S3TaskTypeWritePGFile just skip it.
+			 */
+			if (errno == ENOENT)
+				continue;
+
+			ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("could not open file \"%s\": %m", worker_filename)));
+		}
+
+		while ((readBytes = read(worker_file, buffer, sizeof(buffer))) > 0)
+		{
+			if (write(file, buffer, readBytes) != readBytes)
+				ereport(ERROR,
+						(errcode_for_file_access(),
+						 errmsg("could not write file \"%s\": %m", PGFILES_CRC_FILENAME)));
+		}
+
+		if (readBytes < 0)
+			ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("could not read file \"%s\": %m", worker_filename)));
+
+		close(worker_file);
+	}
+
+	if (pg_fsync(file) != 0)
+		ereport(FATAL,
+				(errcode_for_file_access(),
+				 errmsg("could not fsync file \"%s\": %m", PGFILES_CRC_FILENAME)));
+
+	close(file);
+
+	/* The compaction is completed, now we can remove worker files */
+	for (int i = 0; i < s3_num_workers; i++)
+	{
+		char		worker_filename[MAXPGPATH];
+
+		snprintf(worker_filename, sizeof(worker_filename), "%s.%d",
+				 PGFILES_CRC_FILENAME, i);
+
+		unlink(worker_filename);
+	}
+}
+
+static S3FileChecksum *
+get_worker_pgfiles(void)
```

Comment:
One-line function called only once. Code could be inlined in the caller.

Reply:
I think having a separate function improves readability, hides pointer arithmetic and adds possibility to use it later if necessary.